# Grab bag of accumulated framework patches, hopefully upstreamed
# in the future
diff --git a/easybuild/framework/easyblock.py b/easybuild/framework/easyblock.py
index 4d9b4fbdd..46f174972 100644
--- a/easybuild/framework/easyblock.py
+++ b/easybuild/framework/easyblock.py
@@ -1372,6 +1372,10 @@ class EasyBlock(object):
         if tc_mod in excluded_deps and tc_dep_mods:
             excluded_deps.extend(tc_dep_mods)
 
+        # Compute Canada specific: exclude gcccore/cudacore from toolchain deps, we treat it as build dependency
+        excluded_deps.extend([d for d in tc_dep_mods if d.startswith('gcccore/')])
+        excluded_deps.extend([d for d in tc_dep_mods if d.startswith('cudacore/')])
+
         self.log.debug("List of excluded deps: %s", excluded_deps)
 
         # expand toolchain into toolchain components if desired
@@ -3169,7 +3173,7 @@ class EasyBlock(object):
             raise EasyBuildError("Path to 'env' command to use in shebang lines does not exist: %s", env_for_shebang)
 
         for lang in ['bash', 'perl', 'python']:
-            shebang_regex = re.compile(r'^#![ ]*.*[/ ]%s.*' % lang)
+            shebang_regex = re.compile(r'^#![ ]*.*[/ ]%s' % lang)
             fix_shebang_for = self.cfg['fix_%s_shebang_for' % lang]
             if fix_shebang_for:
                 if isinstance(fix_shebang_for, str):
@@ -4278,7 +4282,7 @@ class EasyBlock(object):
             raise StopException(step)
 
     @staticmethod
-    def get_steps(run_test_cases=True, iteration_count=1):
+    def get_steps(run_test_cases=True, iteration_count=1, multi_deps_extensions_only=False):
         """Return a list of all steps to be performed."""
 
         def get_step(tag, descr, substeps, skippable, initial=True):
@@ -4293,11 +4297,24 @@ class EasyBlock(object):
             (True, lambda x: x.reset_env),
             (True, lambda x: x.handle_iterate_opts),
         ]
+        # if multi_deps is used only for extensions, don't make the build directory each iteration
+        if multi_deps_extensions_only:
+            ready_substeps_iterates = [
+                (False, lambda x: x.check_readiness_step),
+                (True, lambda x: x.reset_env),
+                (True, lambda x: x.handle_iterate_opts),
+            ]
+        else:
+            ready_substeps_iterates = ready_substeps
 
         def ready_step_spec(initial):
             """Return ready step specified."""
-            return get_step(READY_STEP, "creating build dir, resetting environment", ready_substeps, False,
-                            initial=initial)
+            if initial:
+                return get_step(READY_STEP, "creating build dir, resetting environment", ready_substeps, False,
+                                initial=initial)
+            else:
+                return get_step(READY_STEP, "creating build dir, resetting environment",
+                                ready_substeps_iterates, False, initial=initial)
 
         install_substeps = [
             (False, lambda x: x.stage_install_step),
@@ -4337,17 +4354,24 @@ class EasyBlock(object):
         # part 2: iterated part, from 2nd iteration onwards
         # repeat core procedure again depending on specified iteration count
         # not all parts of all steps need to be rerun (see e.g., ready, prepare)
-        steps_part2 = [
-            ready_step_spec(False),
-            extract_step_spec,
-            patch_step_spec,
-            prepare_step_spec,
-            configure_step_spec,
-            build_step_spec,
-            test_step_spec,
-            install_step_spec(False),
-            extensions_step_spec,
-        ] * (iteration_count - 1)
+        if multi_deps_extensions_only:
+            steps_part2 = [
+                ready_step_spec(False),
+                prepare_step_spec,
+                extensions_step_spec,
+            ] * (iteration_count - 1)
+        else:
+            steps_part2 = [
+                ready_step_spec(False),
+                extract_step_spec,
+                patch_step_spec,
+                prepare_step_spec,
+                configure_step_spec,
+                build_step_spec,
+                test_step_spec,
+                install_step_spec(False),
+                extensions_step_spec,
+            ] * (iteration_count - 1)
         # part 3: post-iteration part
         steps_part3 = [
             (POSTITER_STEP, 'restore after iterating', [lambda x: x.post_iter_step], False),
@@ -4378,7 +4402,8 @@ class EasyBlock(object):
         if self.cfg['stop'] == 'cfg':
             return True
 
-        steps = self.get_steps(run_test_cases=run_test_cases, iteration_count=self.det_iter_cnt())
+        steps = self.get_steps(run_test_cases=run_test_cases, iteration_count=self.det_iter_cnt(),
+                               multi_deps_extensions_only=self.cfg["multi_deps_extensions_only"])
 
         # figure out how many steps will actually be run (not be skipped)
         step_cnt = 0
diff --git a/easybuild/framework/easyconfig/default.py b/easybuild/framework/easyconfig/default.py
index 66ed417b6..0531735f4 100644
--- a/easybuild/framework/easyconfig/default.py
+++ b/easybuild/framework/easyconfig/default.py
@@ -167,6 +167,7 @@ DEFAULT_CONFIG = {
     'hiddendependencies': [[], "List of dependencies available as hidden modules", DEPENDENCIES],
     'multi_deps': [{}, "Dict of lists of dependency versions over which to iterate", DEPENDENCIES],
     'multi_deps_load_default': [True, "Load module for first version listed in multi_deps by default", DEPENDENCIES],
+    'multi_deps_extensions_only': [False, "Use multi_deps dependencies only to install extensions", DEPENDENCIES],
     'osdependencies': [[], "OS dependencies that should be present on the system", DEPENDENCIES],
     'moddependpaths': [None, "Absolute path(s) to prepend to MODULEPATH before loading dependencies", DEPENDENCIES],
 
diff --git a/easybuild/framework/easyconfig/tweak.py b/easybuild/framework/easyconfig/tweak.py
index 2c34c83a7..3b4fd6b0e 100644
--- a/easybuild/framework/easyconfig/tweak.py
+++ b/easybuild/framework/easyconfig/tweak.py
@@ -394,6 +394,7 @@ def tweak_one(orig_ec, tweaked_ec, tweaks, targetdir=None):
         try:
             # obtain temporary filename
             fd, tmpfn = tempfile.mkstemp()
+            os.chmod(tmpfn, 0o644)
             os.close(fd)
 
             # write easyconfig to temporary file
diff --git a/easybuild/toolchains/linalg/flexiblas.py b/easybuild/toolchains/linalg/flexiblas.py
index 8b874cc76..02709a805 100644
--- a/easybuild/toolchains/linalg/flexiblas.py
+++ b/easybuild/toolchains/linalg/flexiblas.py
@@ -36,6 +36,7 @@ from easybuild.tools.toolchain.linalg import LinAlg
 
 from easybuild.tools.run import run_shell_cmd
 from easybuild.tools.systemtools import get_shared_lib_ext
+from easybuild.toolchains.compiler.inteliccifort import TC_CONSTANT_INTELCOMP
 
 
 TC_CONSTANT_FLEXIBLAS = 'FlexiBLAS'
@@ -79,6 +80,18 @@ class FlexiBLAS(LinAlg):
     LAPACK_INCLUDE_DIR = [os.path.join('include', 'flexiblas')]
     LAPACK_FAMILY = TC_CONSTANT_FLEXIBLAS
 
+    def _set_blas_variables(self):
+        """Fix the BLAS_LIB for Intel"""
+        if self.COMPILER_FAMILY == TC_CONSTANT_INTELCOMP:
+            self.BLAS_LIB = ['flexiblas_intel']
+            self.BLAS_LIB_MT = ['flexiblas_intel']
+        # and for ILP64
+        if self.options.get('i8', None):
+            # ilp64/i8
+            self.BLAS_LIB = [s.replace('flexiblas', 'flexiblas64') for s in self.BLAS_LIB]
+            self.BLAS_LIB_MT = [s.replace('flexiblas', 'flexiblas64') for s in self.BLAS_LIB_MT]
+        super(FlexiBLAS, self)._set_blas_variables()
+
     def banned_linked_shared_libs(self):
         """
         List of shared libraries (names, file names, paths) which are
diff --git a/easybuild/toolchains/linalg/scalapack.py b/easybuild/toolchains/linalg/scalapack.py
index b4c12e696..666a11c95 100644
--- a/easybuild/toolchains/linalg/scalapack.py
+++ b/easybuild/toolchains/linalg/scalapack.py
@@ -48,3 +48,10 @@ class ScaLAPACK(Blacs):
             return LooseVersion(self.get_software_version(self.SCALAPACK_MODULE_NAME)[0]) < LooseVersion("2.0")
         else:
             return super(ScaLAPACK, self).is_required(name)
+
+    def _set_scalapack_variables(self):
+        """Fix the SCALAPACK_LIB for ILP64"""
+        if self.options.get('i8', None):
+            # ilp64/i8
+            self.SCALAPACK_LIB = ['scalapack64']
+        super(ScaLAPACK, self)._set_scalapack_variables()
diff --git a/easybuild/tools/build_log.py b/easybuild/tools/build_log.py
index 044f53faa..9795df90a 100644
--- a/easybuild/tools/build_log.py
+++ b/easybuild/tools/build_log.py
@@ -269,6 +269,7 @@ def init_logging(logfile, logtostdout=False, silent=False, colorize=fancylogger.
 
             # mkstemp returns (fd,filename), fd is from os.open, not regular open!
             fd, logfile = tempfile.mkstemp(suffix='.log', prefix='easybuild-', dir=tmp_logdir)
+            os.chmod(logfile, 0o644)
             os.close(fd)
 
         fancylogger.logToFile(logfile, max_bytes=0)
diff --git a/easybuild/tools/module_generator.py b/easybuild/tools/module_generator.py
index eebafb3c4..7bd3c815c 100644
--- a/easybuild/tools/module_generator.py
+++ b/easybuild/tools/module_generator.py
@@ -527,6 +527,12 @@ class ModuleGenerator(object):
         """
         raise NotImplementedError
 
+    def set_shell_function(self, key, bashStr, cshStr):
+        """
+        Generate set_shell_function statement in modulefile for the given key/value pair.
+        """
+        raise NotImplementedError
+
     def set_as_default(self, module_dir_path, module_version, mod_symlink_paths=None):
         """
         Set generated module as default module
@@ -1013,6 +1019,12 @@ class ModuleGeneratorTcl(ModuleGenerator):
         # quotes are needed, to ensure smooth working of EBDEVEL* modulefiles
         return 'set-alias\t%s\t\t%s\n' % (key, quote_str(value, tcl=True))
 
+    def set_shell_function(self, key, bashStr, cshStr):
+        """
+        Generate set_shell_function statement in modulefile for the given key/value pair.
+        """
+        raise NotImplementedError
+
     def set_as_default(self, module_dir_path, module_version, mod_symlink_paths=None):
         """
         Create a .version file inside the package module folder in order to set the default version
@@ -1195,6 +1207,8 @@ class ModuleGeneratorLua(ModuleGenerator):
 
         error_msg = 'LmodError("' + error_msg + '")'
         res = self.conditional_statement('userInGroup("%s")' % group, error_msg, negative=True)
+        if 'EASYBUILD_CONFIGFILES' in os.environ:
+            res = '' # disable for Compute Canada:
 
         return res
 
@@ -1469,6 +1483,13 @@ class ModuleGeneratorLua(ModuleGenerator):
         # quotes are needed, to ensure smooth working of EBDEVEL* modulefiles
         return 'set_alias("%s", %s)\n' % (key, quote_str(value))
 
+    def set_shell_function(self, key, bashStr, cshStr):
+        """
+        Generate set_shell_function statement in modulefile for the given key/value pair.
+        """
+        # quotes are needed, to ensure smooth working of EBDEVEL* modulefiles
+        return 'set_shell_function("%s", %s, %s)\n' % (key, quote_str(bashStr), quote_str(cshStr))
+
     def set_as_default(self, module_dir_path, module_version, mod_symlink_paths=None):
         """
         Create a symlink named 'default' inside the package's module folder in order to set the default module version
diff --git a/easybuild/tools/options.py b/easybuild/tools/options.py
index 4068aa1ad..b2b33c562 100644
--- a/easybuild/tools/options.py
+++ b/easybuild/tools/options.py
@@ -2075,6 +2075,7 @@ def set_tmpdir(tmpdir=None, raise_error=False):
             raise EasyBuildError("Failed to create path to temporary directory %s: %s", current_tmpdir, err)
 
     _log.info("Temporary directory used in this EasyBuild run: %s" % current_tmpdir)
+    os.chmod(current_tmpdir, 0o755)
 
     for var in ['TMPDIR', 'TEMP', 'TMP']:
         env.setvar(var, current_tmpdir, verbose=False)
diff --git a/easybuild/tools/robot.py b/easybuild/tools/robot.py
index 1d0546de8..52b19607d 100644
--- a/easybuild/tools/robot.py
+++ b/easybuild/tools/robot.py
@@ -441,7 +441,7 @@ def resolve_dependencies(easyconfigs, modtool, retain_all_deps=False, raise_erro
                         processed_ecs = process_easyconfig(path, validate=not retain_all_deps, hidden=hidden)
 
                         # ensure that selected easyconfig provides required dependency
-                        verify_easyconfig_filename(path, cand_dep, parsed_ec=processed_ecs)
+                        #verify_easyconfig_filename(path, cand_dep, parsed_ec=processed_ecs)
 
                         for ec in processed_ecs:
                             if ec not in easyconfigs + additional:
diff --git a/easybuild/tools/toolchain/linalg.py b/easybuild/tools/toolchain/linalg.py
index bbf158d9e..3ff98576e 100644
--- a/easybuild/tools/toolchain/linalg.py
+++ b/easybuild/tools/toolchain/linalg.py
@@ -92,7 +92,7 @@ class LinAlg(Toolchain):
         # TODO is link order fully preserved with this order ?
         self._set_blas_variables()
         self._set_lapack_variables()
-        if getattr(self, 'MPI_MODULE_NAME', None):
+        if getattr(self, 'MPI_MODULE_NAME', None) and self.SCALAPACK_LIB is not None:
             self._set_blacs_variables()
             self._set_scalapack_variables()
 
